### PR TITLE
chore(cli): add more capabilities to the hotswap CFN evaluate sub-system

### PR DESCRIPTION
**Pull request by skinny85**
The current functionality we use for evaluating CloudFormation in the hotswap part of the CLI is very limited:
only allows substituting the values of parameters used for Assets.
That's not good enough when doing substitutions for StepFunctions State Machines from
[this PR](https://github.com/aws/aws-cdk/pull/16489), for example.

Enhance the capabilities of the CFN eval sub-system by introducing a new class,
`CloudFormationExecutableTemplate`, that allows resolving references to resources inside the template.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+          ret[key] = await this.evaluateCfnExpression(val);
+        }
+        return ret;
+      }
+    }
+
+    return cfnExpression;
+  }
+
+  async 'Fn::Join'(separator: string, args: any[]): Promise<string> {
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:67]: 💅  I'm not super down with these dispatch handlers just living among all the other normal methods on this class. I'd prefer for them to be in a different class, or object, or *something* to distinguish them from the rest of the code here.

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+  readonly parameters: { [parameterName: string]: string };
+  readonly account: string;
+  readonly region: string;
+  readonly partition: string;
+  readonly urlSuffix: string;
+
+  readonly listStackResources: ListStackResources;
+}
+
+export class CloudFormationExecutableTemplate {
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:16]: 💅 Why *executable* ? Why not something like `EvaluateCloudFormationTemplate`? Or if you must have a noun: `CloudFormationTemplateEvaluator` ?

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    const physicalId = resource.PhysicalResourceId;
+
+    // no attribute means Ref expression, for which we use the physical ID directly
+    if (!attribute) {
+      return physicalId;
+    }
+
+    const resourceTypeFormats = RESOURCE_TYPE_ATTRIBUTES_FORMATS[resource.ResourceType];
+    if (!resourceTypeFormats) {
+      return physicalId;
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:168]: Isn't this almost always likely to be wrong? Aren't we better off throwing an error?

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    }
+    const attributeFmtFunc = resourceTypeFormats[attribute];
+    if (!attributeFmtFunc) {
+      return physicalId;
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:172]: And same here?

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    }
+    const service = this.getServiceOfResource(resource);
+    const resourceTypeArnPart = this.getResourceTypeArnPartOfResource(resource);
+    return attributeFmtFunc({
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:176]: Oh my god the incredible hax!

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+interface ArnParts {
+  readonly partition: string;
+  readonly service: string;
+  readonly region: string;
+  readonly account: string;
+  readonly resourceType: string;
+  readonly resourceName: string;
+}
+
+const RESOURCE_TYPE_ATTRIBUTES_FORMATS: { [type: string]: { [attribute: string]: (parts: ArnParts) => string } } = {
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:204]: More tables to maintain ☹️ 

Unfortunately no way around it I suppose... but this is starting to feel dangerous, is it not?

What does SAM CLI do here?

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+  }
+
+  async 'Fn::Sub'(template: string, explicitPlaceholders?: { [variable: string]: string }): Promise<string> {
+    const placeholders = explicitPlaceholders
+      ? { ...this.context, ...(await this.evaluateCfnExpression(explicitPlaceholders)) }
+      : this.context;
+
+    return template.replace(/\${([^}]*)}/g, (_: string, key: string) => {
+      if (key in placeholders) {
+        return placeholders[key];
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:108]: I'm pretty sure this implement is wrong -- already was wrong, nothing to do with you.

Additional lookups that I believe are always valid for `Fn::Sub` are:

- `${LogicalId}` -- evaluate `LogicalId` as a `{ Ref }` if it's not in the context object
- `${LogicalId.Attr}` -- evaluate as a `{ Fn::GetAtt }`.

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts (hunk 1 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:67]: Yep. I didn't like it either, but laziness got the better of me. Fixed it now.

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts (hunk 2 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:16]: Sure, changed to `EvaluateCloudFormationTemplate`.

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts (hunk 3 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:168]: Hmm, really? This is like the mildest problem I see personally with this PR 😃.

What is your concern here?

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts (hunk 5 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:176]: Hmm, really? This is like the mildest problem I see personally with this PR 😃.

What is your concern here?

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts (hunk 3 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:168]: Yep, I think you're right. This change has some pretty far-reaching consequences, but I think it's the correct one to make. Let me know if you like this approach.

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts (hunk 4 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:172]: Same answer as https://github.com/aws/aws-cdk/pull/16696#discussion_r719034348.

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts (hunk 6 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:204]: > Unfortunately no way around it I suppose... but this is starting to feel dangerous, is it not?

A little bit, but this should be fairly stable, so I'm not _that_ worried about it.

> What does SAM CLI do here?

I believe SAM CLI punts this completely to you, and forces you to evaluate these yourself.

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts (hunk 7 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts:108]: Yep, done.

**Comment by skinny85** [comment]
@rix0rrr thanks for the review! I've incorporated all of your comments, would appreciate another pass!

File: packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts (hunk 5 above)
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts]: I misunderstood a little what was happening here. It's... okay :)

File: packages/aws-cdk/lib/api/hotswap/evaluate-cloudformation-template.ts [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+        }
+      }
+
+      async 'Fn::GetAtt'(logicalId: string, attributeName: string): Promise<string> {
+        // ToDo handle the 'logicalId.attributeName' form of Fn::GetAtt
+        const attrValue = await self.findGetAttTarget(logicalId, attributeName);
+        if (attrValue) {
+          return attrValue;
+        } else {
+          throw new CfnEvaluationException(`Attribute '${attributeName}' of resource '${logicalId}' could not be found for evaluation`);
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/hotswap/evaluate-cloudformation-template.ts:88]: This error is somewhat conflating the cases: 

1. No resource with that logical ID
2. We have no idea how to get that particular attribute from this particular type of resource

One is a mistake by the user, the other one is a deficiency in the system that needs to be fixed. Would be great to be able to tell them apart.

Why not throw the error at the site of `return undefined` ?

File: packages/aws-cdk/lib/api/hotswap/evaluate-cloudformation-template.ts [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+        }
+      }
+
+      async 'Fn::Sub'(template: string, explicitPlaceholders?: { [variable: string]: string }): Promise<string> {
+        const placeholders = explicitPlaceholders
+          ? await self.evaluateCfnExpression(explicitPlaceholders)
+          : {};
+
+        // this code is a little convoluted, because we need to deal with async operations
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/hotswap/evaluate-cloudformation-template.ts:97]: Wouldn't it be nicer to define a helper:

```ts
async function asyncRegexReplace(s: string, regex: RegExp, cb: (x: string) => Promise<string>) {
  if (!regex.global) { throw new Error('Regex must be created with /g flag'); }
  const ret = new Array<string>();

  let start = 0;  
  while (true) {
    const m = regex.exec(s);
    if (!m) { break; }

    ret.push(s.substring(start, m.index));
    ret.push(await cb(m[1]));
    start = regex.lastIndex;
  }

  ret.push(s.substr(start));
  return ret.join('');
 
}

**Review by rix0rrr — APPROVED** [review]
Provisional okeedoke, do with my remaining comments as you will

File: packages/aws-cdk/lib/api/hotswap/evaluate-cloudformation-template.ts (hunk 8 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/hotswap/evaluate-cloudformation-template.ts:88]: Yep, you are right. Corrected.

File: packages/aws-cdk/lib/api/hotswap/evaluate-cloudformation-template.ts (hunk 9 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/hotswap/evaluate-cloudformation-template.ts:97]: Oh wow. Thats kind of nuts. I had no idea about this mutability of regular expressions in JavaScript. That's actually kind of scary 😜.

Changed!

**Pull request merged**